altair/vegalite/v5/schema/channels.py

```diff
@@ -38,9 +38,14 @@ def to_dict(self, validate=True, ignore=(), context=None):
                 parsed.pop('type', None)
             elif not (type_in_shorthand or type_defined_explicitly):
                 if isinstance(context.get('data', None), pd.DataFrame):
```

Comment:
As you correctly observed, the `isinstance` condition on line 40 is insufficient as it will not filter out errors related to defined transforms. 
I think it is better to catch this separately or remove this check completely. If you decide to go for the latter than you would have a single `raise ValueError` within `elif not (type_in_shorthand or type_defined_explicitly):`, maybe something as such:

```python
raise ValueError(
    "Error: Unable to determine data type for field '{}'. Please verify that the column"
    " or field name is spelled correctly and that the data type is specified correctly,"
    " particularly if the field is not from a pandas.DataFrame.".format(shorthand)
)
```

Reply:
I like the way you phrased the error message and updated the PR with a similar phrasing, thanks! I would still prefer to have a separate error message for data that is not a pandas data frame since we in those cases know that the issue is a missing data type since the data can't be inferred (although if we remove the other change you commented on than it could also be a misspelled data type I guess, but never a typo due to #2825).